Fix source code validation with windows line ending

academo opened the pull request:
Closes https://github.com/grafana/plugin-validator/issues/205

Fixes an issue where the validator will fail to compare files that were saved in windows against files that were saved on linux.

This occurs when the archive was created with line endings that were different on linux and windows.

There's no risk on having a false positive or no detecting changes because js files are normalized to unix line endings or windows line endings before comparison
but the content stays the same.

academo committed:
Make sure all file line ends are normalized

academo committed:
Validate hash using windows line endings too

academo commented on pkg/analysis/passes/gomanifest/gomanifest.go:
```diff
@@ -7,7 +7,6 @@ import (
 	"encoding/json"
```
notice I applied long line formatting to this file

academo commented on pkg/analysis/passes/gomanifest/gomanifest.go:
```diff
@@ -141,18 +146,22 @@ func verifyManifest(manifest map[string]string, goFiles []string, sourceCodeDir
 		}
 		// check if the sha256sum in the manifest matches the calculated sha256sum
-		if sha256sum != manifestSha256sum {
+		if sha256sum != manifestSha256sum && windowsSha256Sum != manifestSha256sum {
```
this is the fundamental change in the validation of the go manifest. now we calculate two hashes, one with linux line endings and other with windows line endings. We accept either as a valid hash.

marefr replied:
Should it be or instead of and?

academo replied:
This is the condition to fail so passing one is a success (hence and `and`). If we would put an `or` here it'll always fail:

A file is hashed and written into the manifest. the hash is calculated based on the content that might be windows line end or linux line end.

We calculate both hashes of the original file. one with linux line ending and other with windows line ending.

if neither of those hashes matches the hash in the manifest we fail the check.

academo commented on pkg/analysis/passes/gomanifest/gomanifest.go:
```diff
@@ -167,43 +176,23 @@ func verifyManifest(manifest map[string]string, goFiles []string, sourceCodeDir
+// we hash the file content using the sha256 algorithm
+// a second hash is created using the line endings that windows uses
+// for plugins compiled with native windows that use windows line endings
+func hashFileContent(path string) (string, string, error) {
```
because I now have to essentially run replace in the whole content I went with reading the whole file instead of chunks. It is very unlikely we'll have go files big enough to overflow the memory

academo commented on pkg/difftool/difftool.go:
```diff
@@ -65,7 +70,19 @@ func CompareSourceMapToSourceCode(pluginID string, sourceCodeMapFile string, sou
 
-		delta := godiffpatch.GeneratePatch(sourceMapFileName, string(sourceCodeFileContent), sourceMapContent)
+		// Normalize line endings to Unix-style (LF) to ensure consistent comparison
+		cleanSourceCodeFileContent := strings.ReplaceAll(
```
I am normalizing the line ending here before calculating the hash.

marefr commented on pkg/analysis/passes/gomanifest/gomanifest.go:
```diff
@@ -167,43 +176,23 @@ func verifyManifest(manifest map[string]string, goFiles []string, sourceCodeDir
-			fmt.Printf("error closing file for hashing: %v", err)
-		}
-	}()
+	// Calculate the original hash
+	hOrig := sha256.Sum256(data)
+	originalHash := hex.EncodeToString(hOrig[:])
```
Did you consider normalize here as well similar to how you do source map comparison?

academo replied:
the go manifest is a list of the hashed content of the go files at build, the hash is calculated using the line ending of the OS that wrote and we do here is rehash the file and verify. Here we are basically detecting which line ending was used by calculating both hashes and checking.

marefr reviewed (changes requested):
Works/good enough. Added a question. Would be good with some tests as well

academo committed:
Improve code for easier reading

academo committed:
Add tests for windows line endings

academo commented:
> Works/good enough. Added a question. Would be good with some tests as well

@marefr I improved the code variables and legibility to make the intention more clear and added a test case for the windows line ending.

academo committed:
add manifest file to commit

marefr reviewed (approved):
LGTM
